components/application-mgt/org.wso2.carbon.identity.application.mgt/src/main/java/org/wso2/carbon/identity/application/mgt/ApplicationManagementServiceImpl.java

```diff
@@ -852,6 +838,91 @@ public ServiceProvider getServiceProviderByClientId(String clientId, String clie
+                    " tenant: %s", fileName, tenantDomain));
+        }
+
+
+        Collection<ApplicationMgtListener> listeners =
+                ApplicationMgtListenerServiceComponent.getApplicationMgtListeners();
+
+        ServiceProvider savedSP = null;
+        try {
+            if (isUpdate) {
+                savedSP = getApplicationExcludingFileBasedSPs(serviceProvider.getApplicationName(), tenantDomain);
+                if (savedSP == null) {
+                    if (log.isDebugEnabled()) {
+                        log.debug(String.format("Service provider %s@%s is not found, so creating new Service provider",
+                                serviceProvider.getApplicationName(), tenantDomain));
+                    }
+                    isUpdate = false;
+                }
+            }
+
+            if (!isUpdate) {
+                ServiceProvider basicApplication = new ServiceProvider();
+                basicApplication.setApplicationName(serviceProvider.getApplicationName());
+                basicApplication.setDescription(serviceProvider.getDescription());
+                savedSP = addApplication(basicApplication, tenantDomain, username);
+            }
+            serviceProvider.setApplicationID(savedSP.getApplicationID());
+            serviceProvider.setOwner(getUser(tenantDomain, username));
+
+            for (ApplicationMgtListener listener : listeners) {
+                if (listener.isEnable()) {
+                    listener.doPreUpdateApplication(serviceProvider, tenantDomain, username);
+                }
+            }
+            isImportSP.set(true);
+
+            for (ApplicationMgtListener listener : listeners) {
+                if (listener.isEnable()) {
+                    listener.doImportServiceProvider(serviceProvider);
+                }
+            }
+
+            updateApplication(serviceProvider, tenantDomain, username);
+            importerResponse.setApplicationName(serviceProvider.getApplicationName());
+            importerResponse.setErrors(new String[0]);
+            return importerResponse;
+        } catch (IdentityApplicationManagementValidationException e) {
+            importerResponse.setApplicationName(null);
+            importerResponse.setErrors(e.getValidationMsg());
+            if (savedSP != null && !isUpdate) {
+                deleteSavedSP(savedSP, tenantDomain, username);
+            }
+            return importerResponse;
+        } catch (IdentityApplicationManagementException e) {
+            if (savedSP != null && !isUpdate) {
+                deleteSavedSP(savedSP, tenantDomain, username);
+            }
+            String errorMsg = String.format("Error in importing provided service provider %s@%s from file ",
+                    serviceProvider.getApplicationName(), tenantDomain);
+            log.error(errorMsg, e);
+            throw new IdentityApplicationManagementException(errorMsg, e);
```

Comment:
e may not be needed here since this is the service method and e is already logged